optuna/artifacts/_boto3.py

```diff
@@ -0,0 +1,91 @@
+from __future__ import annotations
+
+import io
+import shutil
+from typing import TYPE_CHECKING
+
+import boto3
+from botocore.exceptions import ClientError
+
+from optuna.artifacts.exceptions import ArtifactNotFound
+
+
+if TYPE_CHECKING:
+    from typing import BinaryIO
+
+    from mypy_boto3_s3 import S3Client
+
+
+class Boto3ArtifactStore:
+    """An artifact backend for Boto3.
+
+    Args:
+        bucket_name:
+            The name of the bucket to store artifacts.
+
+        client:
+            A Boto3 client to use for storage operations. If not specified, a new client will
+            be created.
+
+        avoid_buf_copy:
+            If True, skip procedure to copy the content of the source file object to a buffer
+            before uploading it to S3 ins. This is default to False because using upload_fileobj()
+            method of Boto3 client might close the source file object.
+    """
+
+    def __init__(
+        self, bucket_name: str, client: S3Client | None = None, *, avoid_buf_copy: bool = False
+    ) -> None:
+        self.bucket = bucket_name
+        self.client = client or boto3.client("s3")
+        # This flag is added to avoid that upload_fileobj() method of Boto3 client may close the
+        # source file object. See https://github.com/boto/boto3/issues/929.
+        self._avoid_buf_copy = avoid_buf_copy
+
+    def open_reader(self, artifact_id: str) -> BinaryIO:
+        try:
+            obj = self.client.get_object(Bucket=self.bucket, Key=artifact_id)
+        except ClientError as e:
+            if _is_not_found_error(e):
+                raise ArtifactNotFound(
+                    f"Artifact storage with bucket: {self.bucket}, artifact_id: {artifact_id} was"
+                    " not found"
+                ) from e
+            raise
+        body = obj.get("Body")
+        assert body is not None
+        return body
+
+    def write(self, artifact_id: str, content_body: BinaryIO) -> None:
+        fsrc: BinaryIO = content_body
+        if not self._avoid_buf_copy:
+            buf = io.BytesIO()
+            shutil.copyfileobj(content_body, buf)
+            buf.seek(0)
+            fsrc = buf
+        self.client.upload_fileobj(fsrc, self.bucket, artifact_id)
+
+    def remove(self, artifact_id: str) -> None:
+        try:
+            self.client.delete_object(Bucket=self.bucket, Key=artifact_id)
+        except ClientError as e:
+            if _is_not_found_error(e):
+                raise ArtifactNotFound(
+                    f"Artifact storage with bucket: {self.bucket}, artifact_id: {artifact_id} was"
+                    " not found"
+                ) from e
+            raise
```

Comment:
@c-bata 
It seems `client.delete_object` does not raise error when nothing found (See [here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/delete_object.html#:~:text=Removes%20the%20null%20version%20(if%20there%20is%20one)%20of%20an%20object%20and%20inserts%20a%20delete%20marker%2C%20which%20becomes%20the%20latest%20version%20of%20the%20object.%20If%20there%20isn%E2%80%99t%20a%20null%20version%2C%20Amazon%20S3%20does%20not%20remove%20any%20objects%20but%20will%20still%20respond%20that%20the%20command%20was%20successful.)). Shall I just simply refactor like this?
```suggestion
        self.client.delete_object(Bucket=self.bucket, Key=artifact_id)
```

Reply:
@Alnusjaponica Thank you for the delayed response. Good catch. It sounds good 👍

Reply:
@Alnusjaponica It's just an implementation detail that `Boto3Backend.remove()` does not raise an `ArtifactNotFound` exception, so there is no need to make it consistent with other ArtifactStores. If the `ArtifactNotFound` exception is not beneficial for the users, we might need to consider revising the following specification, but let's address it in a follow-up PR.

https://github.com/optuna/optuna/blob/master/optuna/artifacts/_protocol.py#L49-L59

Reply:
I see. Thanks for the comments. I mark this conversation as resesolved.